app/models/metrics/abstract_metrics_factory.py

```diff
@@ -0,0 +1,116 @@
+from abc import ABC # ABC stands for Abstract Base Class
+import logging
+from typing import List, Dict, Optional
+
+import aioboto3
+from aiocache import decorators
+from aws_xray_sdk.core import xray_recorder
+from app.models.metrics.metrics_model import MetricsModel
+
+import app.cache
+import app.config
+
+# batch get has a 100 item limit
+# https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchGetItem.html
+_DYNAMODB_BATCH_GET_ITEM_LIMIT = 100
+
+
+def _chunks(index, n=_DYNAMODB_BATCH_GET_ITEM_LIMIT):
+    """Yield successive n-sized chunks from index."""
+    for i in range(0, len(index), n):
+        yield index[i: i + n]
+
+
+class AbstractMetricsFactory(ABC):
+    _dynamodb_endpoint: str = app.config.dynamodb['endpoint_url']
+    _dynamodb_table: str = None
+    _primary_key_name: str = None
+    _opens_key: str = 'trailing_7_day_opens'
+    _impressions_key: str = 'trailing_7_day_impressions'
+
+    async def get(self, module_id: str, ids: List[str]) -> Dict[str, 'MetricsModel']:
+        """
+        Get engagement metrics for recommendations or slates.
+        - recommendations: The module is a slate that contains the recommendation.
+                           The primary key is <slate id>/<item id>.
+        - slates: The module is the slate lineup that contains the slate.
+                  The primary key is <lineup id>/<slate id>.
+
+        :param module_id: The first part of the primary key
+        :param ids: Used in the second part of the primary key.
+        :return: dictionary of ClickdataModel objects keyed on item (i.e. not including the prefix)
+        """
+        # Keys are namespaced by the module we are getting data from
+        keys = {self._make_key(module_id, i) for i in ids}
+
+        # Always get the default key, it is used for items that don't have any metrics
+        keys.add(self._make_key(module_id, "default"))
+        keys = list(keys)
+
+        metrics = await self._query_cached_metrics(keys)
+        # Remove "/<modules>" suffix and remove None values
+        # TODO: It might be cleaner if this method just returns List[MetricsBaseModel], and callers create the dict
+        # of their choosing.
+        metrics = {k.split("/")[0]: self.parse_from_record(v) for k, v in metrics.items() if v is not None}
+
+        if not metrics:
+            logging.error(f"No metrics for module {module_id} with keys={keys}")
```

Comment:
Why are we logging an error instead of throwing an exception?

Reply:
Answer: because this is currently not unexpected to happen.

If we were to throw an exception, it would be good to create a custom one, so we can trace this event better.